### ISSUE TITLE
Bad default with "Checkout new branch" etc

### ISSUE THREAD

**Issue by elibarzilay** [magit/magit#3351]
I often work on changes and figure out later that I need to stash before
creating a branch for the work.  The confusing thing is that when the
cursor is on the stash the default for the branch start is `stash@{0}`.
The same happens on a `b b`, and probably others.

It's nicely consistent, but probably not too useful, and potentially
confusing for people who don't know enough git.

Is there an option for ignoring refs that have no proper branch name?
Maybe that should be the default for some operations (like `b b` and
`b c`)?

**Comment by tarsius** [comment]
> The confusing thing is that when the cursor is on the stash the default for the branch start is `stash@{0}`.

What command are you invoking for what purpose?

> I often work on changes and figure out later that I need to stash before creating a branch for the work.

It would probably help if you could list all actions that you take to put the above step into context.

> The same happens on a `b b`, and probably others.

At least for this command it makes sense to default to the revision at point - it's the "checkout arbitrary revision" command.

`magit-read-other-branch-or-commit` should probably be changed to not suggest the commit at point (including stashes) if that commit is the current `HEAD`. It does that for the current branch, but not when `HEAD` is detached.

**Comment by elibarzilay** [comment]
1. I'm just using `b c`, when the cursor happens to be on a stash line.

2. I noticed a related confusion: I wanted to switch to master, so I
   quickly entered `b b mas TAB RET` and got an error -- I was already
   *on* `master`, so it was excluded from the completions, leaving it as
   `mas`.  (Disclaimer: I've just upgraded from 2.10.something.)

3. I don't get the "checkout arbitrary revision" comment -- is `b b` now
   able to checkout anything?  Should the new (for me) `b l` be
   preferred for these things?

   * If it is preferred, then maybe include some code to make `b b` do
     the local thing instead?

   * But still that wouldn't make the `b c` situation less confusing...

**Comment by tarsius** [comment]
> I'm just using `b c`, when the cursor happens to be on a stash line.

The part I don't understand is why the cursor is there at that point, which is why I expressed interest in the context.

> I noticed a related confusion: I wanted to switch to master, so I quickly entered `b b mas TAB RET` and got an error -- I was already *on* `master`, so it was excluded from the completions, leaving it as `mas`. (Disclaimer: I've just upgraded from 2.10.something.)

It's possible that I did not exclude this particular invalid choice from the list of choices until after that release. User errors such as the one you describe happen and responding with an error seems like an okay reaction to that. Actually I think it is better to display an unobtrusive error message in this case than to just silently do nothing.

> I don't get the "checkout arbitrary revision" comment -- is `b b` now able to checkout anything?

It was able to do that for a very long time, and it used the thing at point as the default for a long time. That probably had to be a branch at one point to be recognized as a valid default and then it was extended to any (plain) commit and finally got extended to handle the stash special case. Checking out a stash is a valid, though probably not so common, use-case.

> Should the new (for me) `b l` be preferred for these things?

Probably not. That command features some additional "dwim" behavior. I suspect that in your case it would be "do what I didn't want" though. But how would I know, see the documentation for what that command does.

> But still that wouldn't make the `b c` situation less confusing...

That command, and many others, defaulted to using the thing at point for a long time. The only thing that changed is that over time these commands learned to handle certain special cases (i.e. a "weird" thing at point) better.

I don't think making this more consistent is particularly confusing. Defaulting to the thing at point in some cases but not in others, that would be confusing. I can appreciate that in some situations the default choice is not particularly good and that for some use-cases it would be better to fall back to some secondary default, but we cannot in all cases offer the desired default to every user.

**Comment by elibarzilay** [comment]
> The part I don't understand is why the cursor is there at that point,
> which is why I expressed interest in the context.

(I don't remember -- but I'm generally using the cursor as I read stuff,
so it's likely I ended up there when I was visually verifying that the
stash was created...)

> Actually I think it is better to display an unobtrusive error message
> in this case than to just silently do nothing.

I very strongly agree, of course.  (One thing that confused me was the
lack of completion (to `master`), and another is that I'd expect it to
be a magit-level error rather than just running git on my invalid `mas`
which results in an error.)

> That command [`b l`] features some additional "dwim" behavior. I
> suspect that in your case it would be "do what I didn't want"
> though. [...]

Actually, I just tried it, and it looks like it's indeed what I should
use.  (Since I generally use `git` and other tools too, so getting into
a detached head state is something that I really want to avoid.)  So I'm
wondering why you thought that it wouldn't be what I want...?  Are there
any pitfalls there?

(BTW, it looks like using `b l` is even nicer for me than `b c` in that
it accepts a new branch name too, and *then* asks where to start the
branch.  No matter how many time I use `b c`, I'm always entering the
new branch name first and then realize my mistake.)

### BUT...

When I opened this, I was working on a real repo, so I didn't want to
play with dangerous things, and I avoided trying things out since I had
some code mess to deal with.

But now I did.  Originally, I expected something horrible to happen with
a `b c` that starts from a stash...  (Making a branch that starts at the
stash ref, and adding new commits to *that*.)  Now I see that this makes
the new branch start from the commit that the stash was on, and then
pops that stash --- this makes perfect sense.  So what I thought would
result in a mess is actually done very well, so this should probably be
closed.

### Sidenote₁

The only unfortunate thing is that I usually stash before pulling &
intending to create a branch, so this wouldn't be as useful since I'd
still need to update the resulting branch...  Is there a way to make
magit add both `--rebase` and `--autostash` to a pull command?

### Sidenote₂

Maybe this is mid-work, but the info file says that
`magit-branch-pull-request` is on `b R` but its on `b Y`.

**Comment by tarsius** [comment]
> (BTW, it looks like using `b l` is even nicer for me than `b c` in that
> it accepts a new branch name too, and *then* asks where to start the
> branch.  No matter how many time I use `b c`, I'm always entering the
> new branch name first and then realize my mistake.)

You can use `magit-branch-read-upstream-first` to get that order with `b c` too.

**Comment by elibarzilay** [comment]
* `magit-branch-read-upstream-first` is great -- thanks!

* I managed to get what I want with a bad hack -- add

      (magit-define-popup-switch 'magit-pull-popup ?z "Rebase+Autostash" "--rebase --autostash")

  and advice `magit-pull-arguments` to find that bogus flag and split it
  into two flags.  Is there a better way to do that?

[

* As long as I'm asking about it, I also have

      (magit-define-popup-switch 'magit-push-popup ?u "Set upstream" "--set-upstream")

  Because I couldn't get the default value in the magit upstream
  question to be `origin/foo` for a `foo` branch (and that's what a
  plain `git push --set-upstream` does for me with `push.default` set to
  `current`).  Maybe there's something else that I'm missing.

]

**Comment by tarsius** [comment]
> Is there a better way to do that?

* How about just adding two arguments instead? But I guess it is inconvenient having to toggle two arguments. On the other hand I would hope that you can just enable and save `--autostash` once and that that argument would just be ignored when it is not relevant. But I haven't tested that.
* Alternatively you could set `rebase.autoStash`.
* Or going another way, you could create a separate command that hard-codes these arguments. 

> Because I couldn't get the default value in the magit upstream question to be `origin/foo` for a `foo` branch

I would recommend against doing that. The rational is explained at https://magit.vc/manual/magit/The-Two-Remotes.html. If you want to do it anyway, then you could use `b u` to first unset the configured remote and then again to set it to the desired value. But the default choice probably isn't what you would have hoped. Adding the `--set-upstream` argument to the push popup makes sense too.

**Comment by elibarzilay** [comment]
* Yeah, I tried adding two arguments first, then realized that it was
  inconvenient and fragile (in case I forget one).  I even made a
  keyboard macro that toggles both, but that wasn't great either since
  now it required three additions (two flags, and the macro)

* Having an option on by default and doing the config settings are both
  things that I don't want: I want to use this only in specific cases,
  not by default.

* I also tried a quick implementation of a separate command, but then
  figured that this way I'd lose on all of the variants in the existing
  pull, and even getting some of them would require even more hacking.

* Thanks for the remotes pointer, it's a good explanation.  If making
  the `--set-upstream` available makes sense, will you do that change,
  or should I keep my tweak?
